### PR TITLE
basic: add recurse_dir() function as modern replacement for nftw()

### PR DESCRIPTION
**Pull request by poettering**
The new API is built around openat(), O_PATH, statx() and thus be much nicer feature-wise and less prone to races.

### CODE REVIEW

**Comment by poettering** [comment]
Uh, that gcc CI failure is weird. Makes no sense to me at all? Anyone can make sense of this? compiler bug?

File: src/test/test-recurse-dir.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        /* Enumerate the specified dirs in full, once via nftw(), and once via recurse_dir(), and ensure the results are identical */
+        r = recurse_dir_at(AT_FDCWD, p, 0, UINT_MAX, RECURSE_DIR_SORT|RECURSE_DIR_ENSURE_TYPE|RECURSE_DIR_SAME_MOUNT, recurse_dir_callback, &list_recurse_dir);
+        if (r == -ENOENT) {
+                log_warning_errno(r, "Couldn't open directory %s, ignoring: %m", p);
+                return EXIT_TEST_SKIP;
+        }
+
+        assert(r >= 0);
+
+        assert(nftw(p, nftw_cb, 64, FTW_PHYS|FTW_MOUNT) >= 0);
```
Comment (mrc0mmand) [inline src/test/test-recurse-dir.c]: `assert_se()`? Since otherwise it breaks with `-Db_ndebug=true`.

**Comment by mrc0mmand** [comment]
> Uh, that gcc CI failure is weird. Makes no sense to me at all? Anyone can make sense of this? compiler bug?

I guess both the newly introduced tests should use `assert_se()` instead of `assert()` (in cases where we call some function), otherwise the tests might not work as expected with `-Db_ndebug=true`.

**Comment by mrc0mmand** [comment]
This makes the computer sad:

```
 465/1263 test-recurse-dir                                                                            FAIL              0.85s   killed by signal 6 SIGABRT
>>> SYSTEMD_KBD_MODEL_MAP=/build/src/locale/kbd-model-map SYSTEMD_LANGUAGE_FALLBACK_MAP=/build/src/locale/language-fallback-map PATH=/systemd-meson-build:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl MALLOC_PERTURB_=60 /systemd-meson-build/test-recurse-dir
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
entry 17 different: /usr/share/man/da.ISO8859-1/ vs /usr/share/man/da/
Code should not be reached at src/test/test-recurse-dir.c:149, function main(). Aborting. 💥
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

**Comment by mrc0mmand** [comment]
Debian/Ubuntu is still unhappy:

```gdb
Program terminated with signal SIGSEGV, Segmentation fault.
#0  main (argc=1, argv=0x7ffccc8b7ae8) at ../src/src/test/test-recurse-dir.c:148
148	                if (!streq_ptr(list_nftw[i], list_recurse_dir[i])) {
(gdb) bt full
#0  main (argc=1, argv=0x7ffccc8b7ae8) at ../src/src/test/test-recurse-dir.c:148
        i = 0
        list_recurse_dir = 0x0
        p = 0x55780bd962b8 "/usr/share/man"
        r = 0
        __func__ = "main"
        __PRETTY_FUNCTION__ = "main"
(gdb) p list_recurse_dir
$3 = (char **) 0x0
```

because `/usr/share/man` is empty in the mkosi images:

```
# ls -la /usr/share/man
total 0
drwxr-xr-x 1 root root    0 Feb 22  2021 .
drwxr-xr-x 1 root root 1070 Oct  4 18:13 ..
```

reproducer:

```
# mkdir empty
# ./test-recurse-dir empty/
Segmentation fault (core dumped)
```

File: src/basic/recurse-dir.c [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+                bool sx_valid = false;
+                const char *p;
+
+                /* For each directory entry we'll do one of the following:
+                 *
+                 * 1) If the entry refers to a directory, we'll open it as O_DIRECTORY 'subdir' and then statx() the opened directory if requested
+                 * 2) Otherwise and RECURSE_DIR_INODE_FD is set we'll open O_PATH 'inode_fd' and then statx() the opened inode
+                 * 3) Otherwise we'll statx() the directory entry via the directory we are currently looking at
+                 *
+                 */
```
Comment (yuwata) [inline src/basic/recurse-dir.c:125]: Not a matter, but the comment has unnecessary blank line. Did you forget to add something after the enumerated entries??

File: src/basic/recurse-dir.c [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+                                        is_mount = r > 0;
+                                }
+
+                                if (is_mount) {
+                                        r = func(RECURSE_DIR_SKIP_MOUNT,
+                                                 p,
+                                                 dirfd(d),
+                                                 dirfd(subdir),
+                                                 de[i],
+                                                 statx_mask != 0 ? &sx : NULL,
```
Comment (yuwata) [inline src/basic/recurse-dir.c]: Should this be `sx_valid ? &sx : NULL`??

File: src/basic/recurse-dir.c [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+                        if (n_depth_max <= 1) {
+                                /* When we reached maxed depth, generate a special event */
+
+                                r = func(RECURSE_DIR_SKIP_DEPTH,
+                                         p,
+                                         dirfd(d),
+                                         dirfd(subdir),
+                                         de[i],
+                                         statx_mask != 0 ? &sx : NULL,
```
Comment (yuwata) [inline src/basic/recurse-dir.c]: Same here.

File: src/basic/recurse-dir.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * others                        → terminate iteration entirely, return the specified value (idea is that
+ *                                 < 0 indicates errors and > 0 indicates various forms of success)
+ */
+typedef int (*recurse_dir_func_t)(
+                RecurseDirEvent event,
+                const char *path,        /* Full non-normalized path, i.e. the path specified during recurise_dir() with what we found appended */
+                int dir_fd,              /* fd of the current dir */
+                int inode_fd,            /* fd of the current entry in the current dir (O_DIRECTORY if directory, and O_PATH otherwise, but only if RECURSE_DIR_INODE_FD was set) */
+                const struct dirent *de, /* directory entry (always valid) */
+                const struct statx *sx,  /* statx data (unly if statx_mask was non-zero */
```
Comment (yuwata) [inline src/basic/recurse-dir.h]: unly -> only

Also, missing ')'

File: src/basic/recurse-dir.c (hunk 2 above)
Comment (poettering) [reply src/basic/recurse-dir.c:125]: nope, the list is complete, it's the three "clean" cases we have in the inner loop: pin by O_DIRECTORY, pin by O_PATH, and no pinning of the directory entry.

Comment (poettering) [reply src/basic/recurse-dir.c:125]: (will remove the empty line)

File: src/basic/recurse-dir.c (hunk 3 above)
Comment (poettering) [reply src/basic/recurse-dir.c]: no, this is intended: we don#t want to pass the statx data unless the user actually asked for it by setting `statx_mask` to non-zero. We'll query statx() for some other internal reasons, but in that case I dont think we should pass the acquired data along in order not to hide bugs.

Comment (poettering) [reply src/basic/recurse-dir.c]: will add comments about this

File: src/shared/cgroup-setup.c [hunk 6]
```diff
@@ -149,51 +149,59 @@ int cg_blkio_weight_parse(const char *s, uint64_t *ret) {
...
+        r = recurse_dir_at(
+                        AT_FDCWD,
+                        fs,
+                        /* statx_mask= */ 0,
+                        /* n_depth_max= */ UINT_MAX,
+                        RECURSE_DIR_ENSURE_TYPE,
+                        trim_cb,
+                        NULL);
+        if (r < 0)
+                log_debug_errno(r, "Failed to trim cgroup %s: %m", path);
```
Comment (yuwata) [inline src/shared/cgroup-setup.c]: Maybe, add ", ignoring" ??

File: src/shared/cgroup-setup.c [hunk 7]
```diff
@@ -149,51 +149,59 @@ int cg_blkio_weight_parse(const char *s, uint64_t *ret) {
...
+                int q;
+
+                q = cg_hybrid_unified();
                 if (q < 0)
-                        log_warning_errno(q, "Failed to trim compat systemd cgroup %s: %m", path);
+                        return q;
+                if (q > 0) {
+                        q = cg_trim(SYSTEMD_CGROUP_CONTROLLER_LEGACY, path, delete_root);
+                        if (q < 0)
+                                log_debug_errno(q, "Failed to trim compat systemd cgroup %s: %m", path);
```
Comment (yuwata) [inline src/shared/cgroup-setup.c]: Also here. The error will be ignored.

File: src/shared/kbd-util.c [hunk 8]
```diff
@@ -1,94 +1,112 @@
...
         if (!keymap_is_valid(p))
                 return 0;
 
-        r = set_consume(keymaps, TAKE_PTR(p));
-        if (r < 0 && r != -EEXIST) {
-                errno = -r;
-                return -1;
-        }
+        r = set_consume(data->keymaps, TAKE_PTR(p));
+        if (r < 0 && r != -EEXIST)
```
Comment (yuwata) [inline src/shared/kbd-util.c]: Not a matter, but if I remember correctly, `set_consume()` does not return `-EEXIST`.

Also, you can use `set_ensure_consume()`.

File: src/shared/kbd-util.c [hunk 9]
```diff
@@ -117,18 +135,27 @@ int keymap_exists(const char *name) {
...
+                                keymap_recurse_dir_callback,
+                                &(struct recurse_dir_userdata) {
+                                        .keymap_name = name,
+                                });
+                if (r == -ENOENT)
+                        continue;
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to read keymap list from %s, ignoring: %m", dir);
+                        continue;
+                }
```
Comment (yuwata) [inline src/shared/kbd-util.c]: Why do you ignore `-ENOMEM` in this case?

File: src/test/test-kbd-util.c [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+        r = get_keymaps(&maps);
+        if (r < 0) {
+                log_error_errno(r, "Failed to aquire keymaps: %m");
+                return 0;
+        }
+
+        STRV_FOREACH(m, maps) {
+                log_info("Found keympa: %s", *m);
+                assert(keymap_exists(*m) > 0);
```
Comment (yuwata) [inline src/test/test-kbd-util.c]: `assert_se()`?

File: src/test/test-kbd-util.c [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        test_setup_logging(LOG_DEBUG);
+
+        r = get_keymaps(&maps);
+        if (r < 0) {
+                log_error_errno(r, "Failed to aquire keymaps: %m");
+                return 0;
+        }
+
+        STRV_FOREACH(m, maps) {
+                log_info("Found keympa: %s", *m);
```
Comment (yuwata) [inline src/test/test-kbd-util.c]: "keympa" -> "keymap"

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/kbd-util.c (hunk 8 above)
Comment (poettering) [reply src/shared/kbd-util.c]: Hmm, using set_ensure_consume() would mean we couldn't pass a compound struct as userdata anymore, and then couldn't just free the set vi `_cleanup_(set_free_freep)` anymore there, we#d have to define a new destructor to destroy the structure then. So I figured it's simpler to just pre-allocate the Set and pass a pointer to that.

Comment (yuwata) [reply src/shared/kbd-util.c]: Indeed.

File: src/shared/cgroup-setup.c [hunk 12]
```diff
@@ -149,51 +149,59 @@ int cg_blkio_weight_parse(const char *s, uint64_t *ret) {
...
+        r = recurse_dir_at(
+                        AT_FDCWD,
+                        fs,
+                        /* statx_mask= */ 0,
+                        /* n_depth_max= */ UINT_MAX,
+                        RECURSE_DIR_ENSURE_TYPE,
+                        trim_cb,
+                        NULL);
+        if (r < 0)
+                log_debug_errno(r, "Failed to trim cgroup %s, ignoring: %m", path);
```
Comment (yuwata) [inline src/shared/cgroup-setup.c]: Nah, `-ENOENT` should be ignored. Currently, all negative errno is returned at the end.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/cgroup-setup.c [hunk 13]
```diff
@@ -174,27 +178,35 @@ int cg_trim(const char *controller, const char *path, bool delete_root) {
...
+                        fs,
+                        /* statx_mask= */ 0,
+                        /* n_depth_max= */ UINT_MAX,
+                        RECURSE_DIR_ENSURE_TYPE,
+                        trim_cb,
+                        NULL);
+        if (r == -ENOENT) /* non-existing is the ultimate trimming, hence no error */
+                r = 0;
+        else if (r < 0)
+                log_debug_errno(r, "Failed to iterate through cgroup %s, ignoring: %m", path);
```
Comment (yuwata) [inline src/shared/cgroup-setup.c]: Then, the error will not be ignored. Please drop ", ignoring".

File: src/shared/cgroup-setup.c [hunk 14]
```diff
@@ -174,27 +178,35 @@ int cg_trim(const char *controller, const char *path, bool delete_root) {
...
+
+        /* If we shall delete the top-level cgroup, then propagate the faiure to do so (except if it is
+         * already gone anyway). Also, let's debug log about this failure, except if the error code is an
+         * expected one. */
+        if (delete_root && !empty_or_root(path) &&
+            rmdir(fs) < 0 && errno != ENOENT) {
+                if (!IN_SET(errno, ENOTEMPTY, EBUSY))
+                        log_debug_errno(errno, "Failed to trim cgroup %s, ignoring: %m", path);
```
Comment (yuwata) [inline src/shared/cgroup-setup.c]: The error will not be ignored. Please drop ", ignoring".

File: src/shared/cgroup-setup.c [hunk 15]
```diff
@@ -174,27 +178,35 @@ int cg_trim(const char *controller, const char *path, bool delete_root) {
...
 
         q = cg_hybrid_unified();
         if (q < 0)
                 return q;
-        if (q > 0 && streq(controller, SYSTEMD_CGROUP_CONTROLLER)) {
-                q = cg_trim(SYSTEMD_CGROUP_CONTROLLER_LEGACY, path, delete_root);
-                if (q < 0)
-                        log_warning_errno(q, "Failed to trim compat systemd cgroup %s: %m", path);
-        }
+        if (q > 0 && streq(controller, SYSTEMD_CGROUP_CONTROLLER))
```
Comment (yuwata) [inline src/shared/cgroup-setup.c:209]: `controller` is non-NULL? Then, I'd like to add assert() at the beginning. Or, streq_ptr() is better.

**Pull request merged**